File: pgm.lua
local ffi = require "ffi"

ffi.cdef[[
typedef void FILE;
struct pgm_image {
    char magic[2];
    int width, height;
	int max;
    char data[];
};

int pgm_read(FILE *stream, struct pgm_image **out);
int pgm_write(FILE *stream, struct pgm_image *out);
void pgm_free(struct pgm_image *out);

void *calloc(size_t nelem, size_t size);
]]

local cpgm = ffi.load("./libpgm.so")
local pgm = { }

function pgm.create(w,h)

    local img = ffi.cast("struct pgm_image*", ffi.C.calloc(1, ffi.sizeof("struct pgm_image") + w * h))
    img.magic = 'P5'
    img.width = w
    img.height = h
    img.max = 255

    return img

end

function pgm.read(filename)

    local val = ffi.new("struct pgm_image*[1]");
    local rc = cpgm.pgm_read(io.open(filename), val)
    if rc ~= 0 then
        return nil
    end
    val[0] = ffi.gc(val[0], cpgm.pgm_free);

    return val[0][0]

end

function pgm.write(filename,pgm)

    local rc = cpgm.pgm_write(io.open(filename, "w"), pgm)
    return rc == 0

end

return pgm

File: pgm.c
#include <stdio.h>
#include <stdlib.h>

#include "pgm.h"

int pgm_read(FILE *stream, struct pgm_image **out)
{
    char buf[64];

    char magic[2];
    if (!fgets(buf, sizeof buf, stream))
        return 1;
    if (buf[0] != 'P' || buf[1] != '5' || buf[2] != '\n')
        return 2;
    magic[0] = buf[0];
    magic[1] = buf[1];

    char *next;
    if (!fgets(buf, sizeof buf, stream))
        return 3;
    int width = strtol(buf, &next, 10);
    if (!width)
        return 4;
    int height = strtol(next, NULL, 10);
    if (!height)
        return 5;

    if (!fgets(buf, sizeof buf, stream))
        return 6;
    int max = strtol(buf, NULL, 10);
    if (!max)
        return 7;

    struct pgm_image *pgm = malloc(width * height + sizeof *pgm);
    pgm->magic[0] = magic[0];
    pgm->magic[1] = magic[1];

    pgm->width = width;
    pgm->height = height;

    pgm->max = max;

    if (fread(pgm->data, width * height, 1, stream) != 1) {
        free(pgm);
        return 8;
    }

    *out = pgm;

    return 0;
}

int pgm_write(FILE *stream, struct pgm_image *out)
{
    fprintf(stream,
            "%c%c\n"
            "%d %d\n"
            "%d\n"
            , out->magic[0], out->magic[1]
            , out->width, out->height
            , out->max);

    if (fwrite(out->data, out->width * out->height, 1, stream) != 1)
        return 1;

    return 0;
}

void pgm_free(struct pgm_image *out)
{
    free(out);
}

